### PR TITLE
Jm 160744363 add supplier id invoice generator

### PR DESCRIPTION
**Pull request by jacquelineIO**
## Description

This PR adds the `SupplierID` for the Accepted TransportationServiceProvider to the EDI generator. The TransportationServerProvider is accessed via `Shipment > ShipmentOffer > TransportationServiceProviderPerformance > TransportationServiceProvider`. `SupplierID` is a required field for the generation of EDIs before sending to GEX/Syncada.

## Reviewer Notes

`"ShipmentOffers.TransportationServiceProvider"` was removed from the Eager fetch for shipment. It had been recently added and since it has been added I have learned that we want to use `"ShipmentOffers.TransportationServiceProviderPerformance.TransportationServiceProvider"` instead. More [changes](https://www.pivotaltracker.com/story/show/160303967) down the pipeline are planned to remove the use of `"ShipmentOffers.TransportationServiceProvider"`.

### Option 1

`make db_populate_e2e`
Go to tsp-local, login as tsp.
Go to In Transit shipments, click on shipment with locator ENTDEL.
Choose an actual pickup date, and then click on Delivered.
`go run cmd/generate_shipment_edi/main.go --moveID 135af727-f570-4c7e-bf5b-878d717ef83c`

### Option 2

You will have to follow the setup to create a HHG SM move and progress through until the shipment is delivered in order to view the invoice.

### Some specific test to run

The specific `server_test` for this change is:
```shell
go test ./pkg/edi/invoice -testify.m TestEDIString
go test ./pkg/edi/invoice -testify.m TestInvoiceSuite
```
## Code Review Verification Steps

* [ ] Code follows the guidelines for [Logging](./docs/backend.md#logging)
* [ ] The requirements listed in
 [Querying the Database Safely](./docs/backend.md#querying-the-database-safely)
 have been satisfied.
* Any new migrations/schema changes:
  * [ ] Follow our guidelines for zero-downtime deploys (see [Zero-Downtime Deploys](./docs/database.md#zero-downtime-migrations))
  * [ ] Have been communicated to #dp3-engineering
* [ ] There are no aXe warnings for UI.
* [ ] This works in IE.
* Any new client dependencies (Google Analytics, hosted libraries, CDNs, etc) have been:
  * [ ] Communicated to @willowbl00
  * [ ] Added to the list of [network dependencies](https://github.com/transcom/mymove#client-network-dependencies)
* [ ] Tested in the Experimental environment (for changes to containers, app startup, or connection to data stores)
* [ ] Request review from a member of a different team.
* [ ] Have the Pivotal acceptance criteria been met for this change?

## References

* [Pivotal story](https://www.pivotaltracker.com/story/show/160744363) for this change

## Screenshots

None.

### CODE REVIEW

File: pkg/edi/invoice/generator.go [hunk 1]
```diff
@@ -226,29 +225,33 @@ func getHeadingSegments(shipmentWithCost rateengine.CostByShipment, sequenceNum
...
-		return nil, errors.New("No accepted shipment offer found")
-	} else if numAcceptedOffers > 1 {
-		return nil, errors.Errorf("Found %d accepted shipment offers", numAcceptedOffers)
+	acceptedOffer, err := shipment.GetAcceptedShipmentOffer()
+	if err != nil || acceptedOffer == nil {
+		return segments, errors.Wrap(err, "Error retrieving ACCEPTED ShipmentOffer for EDI generator")
+	}
+
+	scac := acceptedOffer.TransportationServiceProviderPerformance.TransportationServiceProvider.StandardCarrierAlphaCode
+	if scac == "" {
```
Comment (kimallen) [inline pkg/edi/invoice/generator.go]: I don't know how our DB works here- could the scac have a value of null if there is no value for it?  Or if there is no value, will it always be ""?

File: pkg/edi/invoice/generator.go [hunk 2]
```diff
@@ -226,29 +225,33 @@ func getHeadingSegments(shipmentWithCost rateengine.CostByShipment, sequenceNum
...
 	}
 	netCentiWeight := float64(*weightLbs) / 100 // convert to CW
 
-	acceptedOffers := shipment.ShipmentOffers.Accepted()
-	numAcceptedOffers := len(acceptedOffers)
-	if numAcceptedOffers == 0 {
-		return nil, errors.New("No accepted shipment offer found")
-	} else if numAcceptedOffers > 1 {
-		return nil, errors.Errorf("Found %d accepted shipment offers", numAcceptedOffers)
+	acceptedOffer, err := shipment.GetAcceptedShipmentOffer()
```
Comment (kimallen) [inline pkg/edi/invoice/generator.go]: Nice refactor

File: pkg/models/transportation_service_provider.go [hunk 3]
```diff
@@ -19,6 +19,7 @@ type TransportationServiceProvider struct {
 	StandardCarrierAlphaCode string    `json:"standard_carrier_alpha_code" db:"standard_carrier_alpha_code"`
 	Enrolled                 bool      `json:"enrolled" db:"enrolled"`
 	Name                     *string   `json:"name" db:"name"`
+	SupplierID               *string   `json:"supplier_id" db:"supplier_id"`
```
Comment (kimallen) [inline pkg/models/transportation_service_provider.go]: I'm confused as to how my table (before creating a db from your PR) already had a SupplierID field if the SupplierID didn't already exist on the model.

File: pkg/edi/invoice/generator.go (hunk 1 above)
Comment (kimallen) [reply pkg/edi/invoice/generator.go]: I think I answered my own question- a string can't be nil in golang. (unless it's a pointer to a string)

File: pkg/models/transportation_service_provider.go (hunk 3 above)
Comment (jacquelineIO) [reply pkg/models/transportation_service_provider.go]: @kimallen the field was already in the database, there was just no `go` code to handle it. And it was being hard-coded into the EDI so it was not missed.

File: pkg/edi/invoice/generator_test.go [hunk 4]
```diff
@@ -102,13 +102,16 @@ func helperCostsByShipment(suite *InvoiceSuite) []rateengine.CostByShipment {
...
+	supplierID := scac + "1234" //scac + payee code -- ABCD1234
 	shipmentOffer := testdatagen.MakeShipmentOffer(suite.db, testdatagen.Assertions{
 		ShipmentOffer: models.ShipmentOffer{
 			Shipment: shipment,
 			Accepted: swag.Bool(true),
 		},
 		TransportationServiceProvider: models.TransportationServiceProvider{
-			StandardCarrierAlphaCode: "ABCD",
+			StandardCarrierAlphaCode: scac,
+			SupplierID:               &supplierID,
```
Comment (akostibas) [inline pkg/edi/invoice/generator_test.go]: Maybe it would be worthwhile to beef up the Test Data Generator so that it creates random SupplierIDs by default?
https://github.com/transcom/mymove/blob/master/pkg/testdatagen/make_tsp_records.go#L34

**Comment by Ryan-Koch** [comment]
Looks good to me!

File: pkg/testdatagen/make_tsp_records.go [hunk 5]
```diff
@@ -28,12 +29,32 @@ func RandomSCAC() string {
...
+		if err := db.Find(&existingTsp, assertions.TransportationServiceProvider.ID); err == nil {
+			// Found existing TSP for this ID
+			return existingTsp
+		}
+	}
+
 	scac := assertions.TransportationServiceProvider.StandardCarrierAlphaCode
 	if scac == "" {
-		scac = RandomSCAC()
+		scac = "ABCD" //RandomSCAC()
```
Comment (akostibas) [inline pkg/testdatagen/make_tsp_records.go]: Why did you remove `RandomSCAC()`? If there's a good reason to remove it, I don't think it's worth keeping the comment of the old function name.

File: pkg/edi/invoice/generator_test.go (hunk 4 above)
Comment (akostibas) [reply pkg/edi/invoice/generator_test.go]: Oh, you did! :) Are you relying on a specific SupplierID here? It doesn't look like it. Maybe you can omit this code.

Comment (jacquelineIO) [reply pkg/edi/invoice/generator_test.go]: @akostibas the problem is the EDI golden file. I'm not sure why this wasn't an issue for anyone else, but when I was testing I started getting random SCACs and it wasn't matching the EDI golden file. So the code is setup that it can handle changing it to random but we'd need to figure out what to do with the golden file. And that's a good conversation to have, because I don't know off-hand how to handle that.

File: pkg/testdatagen/make_tsp_records.go (hunk 5 above)
Comment (jacquelineIO) [reply pkg/testdatagen/make_tsp_records.go]: @akostibas EDI golden file fails in testing

File: pkg/models/shipment.go [hunk 6]
```diff
@@ -782,3 +783,50 @@ func (s *Shipment) FetchShipmentLineItemsByItemID(db *pop.Connection, tariff400n
...
+		s.Status == ShipmentStatusINTRANSIT ||
+		s.Status == ShipmentStatusDELIVERED ||
+		s.Status == ShipmentStatusCOMPLETED {
+		return true
+	}
+	return false
+}
+
+// GetAcceptedShipmentOffer returns the ShipmentOffer for an Accepted TSP for a Shipment
+func (s *Shipment) GetAcceptedShipmentOffer() (*ShipmentOffer, error) {
```
[the remark this answers is not in the document]
Comment (mikena-truss) [reply pkg/models/shipment.go]: Since this is an exported method, it should probably have some unit tests.

File: pkg/models/shipment.go [hunk 7]
```diff
@@ -782,3 +783,50 @@ func (s *Shipment) FetchShipmentLineItemsByItemID(db *pop.Connection, tariff400n
...
+		}
+	} else if numAcceptedOffers == 0 {
+		// If the Shipment does not require that it has a TSP then return nil
+		// -- The Shipment is currently in a state that doesn't require a TSP to be associated to it
+		return nil, nil
+	}
+
+	// Double-check for nil before accessing the variable
+	if acceptedOffers == nil {
+		return nil, nil
```
Comment (mikena-truss) [inline pkg/models/shipment.go]: Should this return an error? Everywhere it is used, it checks for nil then passes an error on.

**Comment by mikena-truss** [comment]
Mostly looks good! I have some style suggestions and I'd like to see some tests on that tsp method :)

On a separate note, I'm noticing in a lot of PRs we're referencing nested associations. I'm curious if you (or anyone else) has thoughts on how we can make that be a little less redundant, since it's hard to catch these in PRs and annoying to write/read.

File: pkg/models/transportation_service_provider.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (mikena-truss) [reply pkg/models/transportation_service_provider.go]: This might not be something to address in the PR btw, I'm just raising it as a concern :)

[the remark this answers is not in the document]
Comment (akostibas) [reply pkg/models/transportation_service_provider.go]: I can understand that concern. If we were using Payee Code anywhere else in the system, I would want to break it out. But as far as we have seen, we use the `SupplierID` as a distinct value and never need to handle the components separately. I believe the Payee code is mainly a US Bank artifact that we don't need to work with independently.

File: pkg/models/shipment.go (hunk 6 above)
[the remark this answers is not in the document]
Comment (jacquelineIO) [reply pkg/models/shipment.go]: @mikena-truss  relying on `go test ./pkg/edi/invoice -testify.m TestEDIString` for the unit test.

[the remark this answers is not in the document]
Comment (mikena-truss) [reply pkg/models/shipment.go]: There should probably be a `TestAcceptedShipmentOffer` where you can test multiple cases for the method. The `TestEDIString` should probably be thought of as more of a integration test where we're just making sure everything ties together nicely in the most ideal case.

File: pkg/models/shipment.go (hunk 7 above)
Comment (jacquelineIO) [reply pkg/models/shipment.go]: No. At this point, It's OK to not have an accepted offer. This means that the shipment is not in a state the one is required.

File: pkg/models/shipment_offer_test.go [hunk 8]
```diff
@@ -79,12 +79,14 @@ func (suite *ModelSuite) TestShipmentOfferStateMachine() {
 func (suite *ModelSuite) TestGetAcceptedShipmentOffer() {
 	// Trying a nil slice of shipment offers.
 	var shipmentOffers ShipmentOffers
-	suite.Nil(shipmentOffers.Accepted())
+	shipmentOffers, _ = shipmentOffers.Accepted()
```
[the remark this answers is not in the document]
Comment (jacquelineIO) [reply pkg/models/shipment_offer_test.go]: Added the additional error checking
